### PR TITLE
Base model

### PR DESCRIPTION
**Pull request by streeve**
Combine functionality shared across models. 

~~@pzhanggit I wasn't exactly sure how to separate the graph layer initialization from the multi-head init, but I think this moves in the direction we discussed. Kind of the same thing, should we have forward propagation functions with and without multi-heads?~~

With this approach, all models now have multi-heading which is tested in the CI (running without error, no specific accuracy checks yet)

### CODE REVIEW

**Comment by pzhanggit** [comment]
> Combine functionality shared across models. @pzhanggit I wasn't exactly sure how to separate the graph layer initialization from the multi-head init, but I think this moves in the direction we discussed. Kind of the same thing, should we have forward propagation functions with and without multi-heads?

@streeve  Not sure either. Could we do something like define a class with GNN structures only, without forward propagation and loss etc.  and then use it as a layer before the multi-head part where we define the model with init, forward and loss?

**Comment by allaffa** [comment]
@streeve do you want me to approve the merge request?

**Review by allaffa — APPROVED** [review]

**Comment by streeve** [comment]
This is not ready to merge - I need to fix some mistakes

**Comment by streeve** [comment]
I updated with new tests across all models after I realized in the initial version I didn't correctly inherit from the new base class for anything but PNN, but it passed anyway since we didn't test the other model types.

File: utils/utils.py
Comment (streeve) [inline utils/utils.py:59]: I'm a concerned about doing this, but it's the only way I could get the tests for all models to run right now

**Comment by streeve** [comment]
@pzhanggit this is ready for review, but won't pass tests at least until #27 is merged

File: utils/utils.py
Comment (streeve) [reply utils/utils.py]: Replaced with correct check for whether DDP has already been initialized

**Comment by streeve** [comment]
Okay @allaffa @pzhanggit this is ready for final review. I'm ready to merge if you approve

**Pull request merged**
